woocommerce/woocommerce
# Opt-in Checkout endpoint to rate limits.

### Submission Review Guidelines:

This PR opts-in the Checkout POST endpoint to rate limits by default behind a flag. The limit here would be 3 requests every 60 seconds. This only changes the default values (hooks in early) so any actual consumer can revert this if needed.

This would be the equivalence of running this code:

https://gist.github.com/senadir/ee4dccbcfd6c125e9e3eabddaa4b2822

You can override this by running this snippet:

```php
<?php
add_filter(
    'woocommerce_store_api_rate_limit_options',
    function ( $options ) {
        $options['enabled'] = false;
        return $options;
    }
);
```

See full docs for rate limiting here

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/StoreApi/docs/rate-limiting.md



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. This is disabled by default, go to WooCommerce -> Settings -> Advanced -> Features and enable Rate Limiting for Checkout.
1. You need to have some sort of flow that will trigger an error on Checkout (incorrect card, creating an account that already exists), as placing an order and coming back will be tiring.
2. Try to place the wrong order once, it will fail with whatever error that should be (account already exists, wrong card...)
3. Try a couple of times more.
4. The forth time, you should get an error telling you that you exceed the limits and need to wait 60s.
5. After 60s, correct the info and place a correct order, it should pass.
6. Disable Rate limitng feature, test again, you should not be rate limited after a third request


### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

Add feature to enable tate limits for Store API's Checkout endpoint and Checkout block.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
